### PR TITLE
removed name locator strategy from android and ios drivers along with tests

### PR DESCRIPTION
**Pull request by SrinivasanTarget**
@TikhomirovSergey please review

### CODE REVIEW

File: src/main/java/io/appium/java_client/DefaultGenericMobileElement.java [hunk 1]
```diff
@@ -100,11 +100,17 @@ public List findElementsByTagName(String using) {
         return super.findElementsByTagName(using);
     }
 
-    public T findElementByName(String using) {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/DefaultGenericMobileElement.java]: Should be reverted

File: src/main/java/io/appium/java_client/DefaultGenericMobileElement.java [hunk 2]
```diff
@@ -100,11 +100,17 @@ public List findElementsByTagName(String using) {
...
+    /**
+     * @throws WebDriverException This method doesn't work against native app UI.
+     */
+    public T findElementByName(String using) throws WebDriverException{
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/DefaultGenericMobileElement.java]: Should be reverted

File: src/main/java/io/appium/java_client/DefaultGenericMobileElement.java [hunk 3]
```diff
@@ -100,11 +100,17 @@ public List findElementsByTagName(String using) {
...
         return (T) super.findElementByName(using);
     }
 
-    public List findElementsByName(String using) {
+    /**
+     * @throws WebDriverException This method doesn't work against native app UI.
+     */
+    public List findElementsByName(String using) throws WebDriverException{
         return super.findElementsByName(using);
     }
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/DefaultGenericMobileElement.java]: Should be reverted

File: src/main/java/io/appium/java_client/pagefactory/AndroidFindBy.java [hunk 4]
```diff
@@ -27,7 +27,7 @@
  * element or a list of elements. Used in conjunction with
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/pagefactory/AndroidFindBy.java]: Should be reverted

File: src/main/java/io/appium/java_client/pagefactory/SelendroidFindBy.java [hunk 5]
```diff
@@ -27,13 +27,12 @@
  * element or a list of elements. Used in conjunction with
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/pagefactory/SelendroidFindBy.java]: Let's keep it for some time

File: src/test/java/io/appium/java_client/pagefactory_tests/SelendroidModeTest.java [hunk 6]
```diff
@@ -50,15 +50,6 @@
     @SelendroidFindBy(id = "text1")
     private WebElement textSelendroidId;
 
-    @SelendroidFindBy(name = "Accessibility")
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/SelendroidModeTest.java]: If Selendroid mode supports by name  then this change should be reverted

File: src/test/java/io/appium/java_client/pagefactory_tests/iOSPageObjectTest.java [hunk 7]
```diff
@@ -118,13 +118,13 @@
 	
 	@iOSFindAll({
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/pagefactory_tests/iOSPageObjectTest.java]: Is this test passing?

File: src/test/java/io/appium/java_client/ios/IOSDriverTest.java [hunk 8]
```diff
@@ -124,7 +124,7 @@ public void hideKeyboardWithParametersTest() {
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/IOSDriverTest.java]: @bootstraponline @imurchie 
By.name is deprecated? iOS automation

File: src/test/java/io/appium/java_client/ios/iOSGestureTest.java [hunk 9]
```diff
@@ -134,7 +134,7 @@ public void TapSingleFingerTest() {
 
   @Test
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/iOSGestureTest.java]: @bootstraponline @imurchie 
By.name is deprecated? iOS automation

File: src/main/java/io/appium/java_client/pagefactory/iOSFindBy.java [hunk 10]
```diff
@@ -27,7 +27,7 @@
  * element or a list of elements. Used in conjunction with
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/pagefactory/iOSFindBy.java]: @bootstraponline @imurchie 
By.name is deprecated? iOS automation

This change should be reverted anyway

File: src/test/java/io/appium/java_client/ios/iOSGestureTest.java (hunk 9 above)
Comment (SrinivasanTarget) [inline src/test/java/io/appium/java_client/ios/iOSGestureTest.java]: @TikhomirovSergey https://github.com/appium/appium-ios-driver/blob/master/lib/driver.js#L75

This is removed from ios driver as well.Please look into this.Correct me if i am wrong.

**Comment by TikhomirovSergey** [comment]
Hi @SrinivasanTarget 
First of all let's add descriptions of PR. Ok?

I'm against these changes as they are because: 
- it is breaking and harmfull for users.

- I'm not sure that tests will pass;
Take a look at these
https://github.com/appium/java-client/pull/313/files#diff-54b126e009942041858b50db621bc3aeR121
https://github.com/appium/java-client/pull/313/files#diff-54b126e009942041858b50db621bc3aeR127

- Also I'm not sure that _By.name_ is deprecated for Selendroid. Also I couldn't check check it because of some current problems with 1.5.x. @bootstraponline @imurchie Is _By.name_ not supported by Selendroid mode. It can be used instead of _accessibility_?

@SrinivasanTarget 
But you can make some improvements and help ne to resolve #311 

What should be done if you want this PR to be merged: 
- revert changes which require reversion. Read my comments. 

- The _name_ option of annotatations should be annotated _Deprecated_. You can do it this way 

```java
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD, ElementType.TYPE})
public @interface AndroidFindBy {
	String uiAutomator() default "";
	String accessibility()  default "";
	String id() default "";
	@Deprecated
	/**
	 * By.name selector is not supported by Appium server node since 1.5.x.
	 * So this option is going to be removed further. Be careful.
	 */
	String name() default "";
	String className() default "";
	String tagName() default "";
	String xpath()  default "";
}
```
So users will be warned and they will have time to improve their projects. We will remove it later.  But I still don't know which _name_ options should be deprecated. @bootstraponline @imurchie is _By.name_ deprecated for Selendroid? One more question. Why _By.name_ is deprecated for iOS? iOS-elements have the _name_ tag, I think.

So the point 
> mark name parameter Decrecated as By.name selector is not supported by native app automation

will be finished.

- You could finish this point  

>org.openqa.selenium.InvalidSelectorException should be handled.

The target code is here: 
https://github.com/appium/java-client/blob/6f7615af626e201bb6431fb79945488c3c0a5743/src/main/java/io/appium/java_client/pagefactory/ThrowableUtil.java#L26. The exception 

 ```java
org.openqa.selenium.InvalidSelectorException: Locator Strategy 'css selector' is not supported for this session (WARNING: The server did not provide any stacktrace information)
```

is thrown and it is not being handled here. We should keep it backward compatible with Appium node 1.4x for some time. So you could improve this method the following way: 
 
    1) firstly it checks the excepion type. If it is _org.openqa.selenium.InvalidSelectorException_ then method should return true. 

   2) if 1) has a negative result then it checks exception message. It could use regexps. If message is convenient to the previous pattern 
https://github.com/appium/java-client/blob/6f7615af626e201bb6431fb79945488c3c0a5743/src/main/java/io/appium/java_client/pagefactory/ThrowableUtil.java#L24

or   
 
 ```java
Locator Strategy 'xxxr' is not supported bla-bla-bla
```

then then method should return true. 

 3) if both 1) and 2) have negative results then these steps are repeated recursively with the cause of exception.

File: src/main/java/io/appium/java_client/DefaultGenericMobileElement.java (hunk 3 above)
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/DefaultGenericMobileElement.java]: ...By.nane is still supported by browser automation tools

**Comment by SrinivasanTarget** [comment]
Thanks @TikhomirovSergey for your time.Will update the descriptions going forward. And will update the PR based on your review comments.

**Comment by bootstraponline** [comment]
> @bootstraponline @imurchie is By.name deprecated for Selendroid?

[No.](https://github.com/selendroid/selendroid/blob/c2ec0d49f18279e751ce0500f2454e7c494657f0/selendroid-server/src/main/java/io/selendroid/server/model/internal/NativeAndroidBySelector.java#L38) Even if it was deprecated (it isn't), we'd still want to wait for support to be removed from the server. If the server supports it then I don't see why clients should remove the feature.

**Comment by SrinivasanTarget** [comment]
Thanks for clarification. Will update my piece of code.

**Comment by TikhomirovSergey** [comment]
@SrinivasanTarget 
Ok. I've seen your changes. They are looking much better now :). I'm going to review and merge them soom.

**Comment by SrinivasanTarget** [comment]
I'm still working on this. will update once it's done.

**Comment by TikhomirovSergey** [comment]
ok

**Comment by SrinivasanTarget** [comment]
@TikhomirovSergey I have handled both the scenario's mentioned above.Please review.

**Comment by TikhomirovSergey** [comment]
@SrinivasanTarget I'm ok. :+1: 
@imurchie?

**Comment by imurchie** [comment]
Conceptually it looks fine. I can't comment on the Java code.

**Pull request merged**